### ISSUE TITLE
Enhancement: [prefer-nullish-coalescing] should allow if statement assignment (??=) to be optional

### ISSUE THREAD

**Issue by asselin** [typescript-eslint/typescript-eslint#10995]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [x] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [x] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/prefer-nullish-coalescing/

### Description

This #10829 recent enhancement makes the rule check and recommend nullish coalescing for if() statements.  IMO, this is _not_ simpler to read, and won't include this to my organization's set of coding rules.  The other recommendations from this rule are useful, and I'd like to keep it enabled, but this rule doesn't have a way to turn off the recommendation for the if() case specifically.

Please add an option to disable nullish coalescing checks for if() statements.  Suggested name: ignoreIfStatements

### Fail

```TypeScript
// This code fails by default with the new version of this rule

if (!foo) {
  foo = callAFunctionToGetAValue();
}
```

### Pass

```TypeScript
// This code should PASS if the new option (suggested name: ignoreIfStatements) is set to TRUE

if (!foo) {
  foo = callAFunctionToGetAValue();
}
```

**Comment by bobber205** [comment]
This is a completely insane change to enforce as a "standard". I legit thought it was a joke at first when a coworker told me. There's no universe where `??=` is easier to read. We cannot upgrade without this becoming optional, hopefully by default.

**Comment by kirkwaiblinger** [comment]
Hi @asselin ! Yep, that's a good point, this should be configurable with an option, like the rest of the checks in this rule. 

@bobber205 I understand mistakes can be frustrating, but please avoid jumping straight to language like calling things "insane". This was nothing but a simple oversight which is easy to rectify. We, meaning us on the team as well as the generous community contributors who pitch and implement features, do our best to get things right the first time. But, sometimes things slip through the cracks, and we are grateful for community feedback such as this issue report to find out about them so that we can fix them.

**Comment by kirkwaiblinger** [comment]
cc @OlivierZal

**Comment by OlivierZal** [comment]
@kirkwaiblinger, I can take it.

**Comment by OlivierZal** [comment]
@kirkwaiblinger, I can take it. 

Do we go with an opt-in or an opt-out? Opt-in is less breaking-change.

**Comment by kirkwaiblinger** [comment]
> opt-in or an opt-out? Opt-in is less breaking-change.

Personally, I'd say just opt-out because 
- we're already here now
- the rest of the options are opt-out

But not strongly opinionated, either way is fine by me 👍

**Comment by kirkwaiblinger** [comment]
On the naming, the proposed `ignoreIfStatements` is [also used in the optional chaining corollary PR](https://github.com/typescript-eslint/typescript-eslint/pull/10137/files#diff-6c3803ce84d2677266655bc6bfb2851b9002e0df33f575381d2b5a0b2d6415e2R268), so I think let's definitely use that name.

**Comment by bradzacher** [comment]
```ts
if (!foo) {
  foo = callAFunctionToGetAValue();
}

// vs

foo ??= callAFunctionToGetAValue();
```

🤷 Looks clearer and cleaner to me but hey it's a stylistic decision I guess.

**Comment by asselin** [comment]
I'm good with opt-out.  I expect when upgrading eslint that there will be issues to deal with, either fixing code, or shutting off new rules / adding options to opt out of new behavior.

Thank you @kirkwaiblinger, @OlivierZal, @JoshuaKGoldberg, @bradzacher for all your work on eslint!

**Comment by asselin** [comment]
> I'm definitely +1 on giving an option for this, but am surprised at the immediate and strong pushback about how undesirable enforcing `??=` over `if` is. Could someone please post a few "before and after" code snippets with an explanation of why they're so stylistically inferior?
> 
> This is not a request to engage angrily or debate, I just want to understand 🙂

You asked for it, now you pay the penalty of reading my ramblings about how I read and understand code :)

When I read:

```
if (!foo) {
  foo = callAFunctionToGetAValue();
}
```

My internal dialog is "If not foo, then foo equals .......", which immediately triggers "this is setting a default".  "??=" is a  shorthand that I still need to mentally expand to understand what it's doing:

```
foo ??= callAFunctionToGetAValue();
```

"Foo question-question-equals callAFunctionToGetAValue.... that's the same as foo equals foo question-question callAFunctionToGetAValue".  Then I have to picture that in my mind:

```
foo = foo ?? callAFunctionToGetAValue();
```

Then go back and start parsing again: "foo equals foo or* callAFunctionToGetAValue. ... ok, that's setting a default".  "or*" is my mental "this is almost or (||) but not exactly... remember that in case it's important"

I think "??" is now getting to a point of general adoption with the vast population of JS/TS devs, and this rule (particularly with autofixes) helps drive it.  But I don't think "??=" is there yet.  Part of the problem may be that there's no good name to use when reading "??" or "??=" (contrast with e.g. "plus-equals" for +=, or "or-equals" for ||=).  I also don't think just recognizing "??=" as a pattern and internalizing it as an idiom is widespread yet.  So I would expect many other devs are going to go through the same mental expansion process, which is of course much slower than speed reading through an if statement.

My 2c

**Comment by bradzacher** [comment]
Nitpick about the equivalence.

```ts
foo ??= callAFunctionToGetAValue();

// equivalent to

foo ?? (foo = callAFunctionToGetAValue());
```

This is an minor distinction from your interpretation - your interpretation always does an assignment, regardless of if `foo` is nullish. Whereas this representation **only** does the assignment if `foo` is nullish. A slight perf optimisation and also ensures that setters aren't triggered unless the value changes.

**Comment by neilenns** [comment]
Please add this, and it should have been part of the original edition and disabled by default. The change started causing failures in my CI build for code that was previously fine.

**Comment by neilenns** [comment]
Please add this, and it should have been part of the original edition and disabled by default. The change started causing failures in my CI build for code that was previously fine.

**Comment by JavaScriptBach** [comment]
FWIW the new behavior of this rule was also polarizing at my company, so having an option to disable it sounds good. However, personally I don't think that the current behavior makes code any stylistically worse, and certainly not significantly worse. If one is willing to use `??`, then it doesn't seem like `??=` requires much additional cognitive overhead. I was not familiar with `??=` before the latest release, but by the time I'd finished fixing the lint errors in my codebase, I felt like it was slightly more readable than `if` statements with nesting.

**Comment by kirkwaiblinger** [comment]
Personally, I like that you only have to specify the assignee once 🤷‍♂ 

```ts
if (foo.bar == null) {
   foo.bat = 'typo, whoops';
}

// vs

foo.bar ??= 'all good';
```

but we're all on the same page that it's totally subjective
